contracts/0.8.9/WithdrawalQueue.sol

```diff
@@ -277,21 +279,79 @@ contract WithdrawalQueue is AccessControlEnumerable, WithdrawalQueueBase {
+        for (uint256 i = 0; i < _withdrawalRequestInputs.length; ++i) {
+            requestIds[i] = _enqueueWithdrawalRequest(msg.sender, _withdrawalRequestInputs[i]);
+        }
+    }
+
+    struct Permit {
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+        address owner;
+        uint256 deadline;
+    }
+
+    /// @notice Request the sequence of withdrawals according to passed `withdrawalRequestInputs` data using EIP-2612 Permit
+    /// @param _withdrawalRequestInputs an array of `WithdrawalRequestInput` data. The standalone withdrawal request will
+    ///  be created for each item in the passed list. If `WithdrawalRequestInput.recipient` is set to `address(0)`,
+    ///  `msg.sender` will be used as recipient.
+    /// @return requestIds an array of the created withdrawal requests
+    function requestWithdrawalsWithPermit(
+        WithdrawalRequestInput[] memory _withdrawalRequestInputs,
+        Permit[] memory _permits
+    ) external whenResumed returns (uint256[] memory requestIds) {
+        if (_withdrawalRequestInputs.length != _permits.length)
+            revert LengthsMismatch(_withdrawalRequestInputs.length, _permits.length);
+        requestIds = new uint256[](_withdrawalRequestInputs.length);
+        for (uint256 i = 0; i < _withdrawalRequestInputs.length; ++i) {
+            IERC20Permit(_withdrawalRequestInputs[i].token).permit(
+                _permits[i].owner,
+                address(this),
+                _withdrawalRequestInputs[i].amount,
+                _permits[i].deadline,
+                _permits[i].v,
+                _permits[i].r,
+                _permits[i].s
+            );
+            requestIds[i] = _enqueueWithdrawalRequest(_permits[i].owner, _withdrawalRequestInputs[i]);
+        }
     }
 
     /// @notice Claim withdrawals batch once finalized (claimable)
-    /// NB: Always reverts
-    function claimWithdrawalBatch(uint256[] calldata /*_requests*/ ) external pure {
-        revert Unimplemented();
+    /// @param _requestIds ids of the finalized requests to claim
+    /// @param _hints rate indices found offchain that should be used for claiming
+    function claimWithdrawals(uint256[] calldata _requestIds, uint256[] calldata _hints) external {
+        if (_requestIds.length != _hints.length) revert LengthsMismatch(_requestIds.length, _hints.length);
+        for (uint256 i = 0; i < _requestIds.length; ++i) {
+            claimWithdrawal(_requestIds[i], _hints[i]);
+        }
+    }
+
+    /// @notice Returns the list of hints for the given request ids
+    /// @param _requestIds ids of the requests to get hints for
+    function findClaimHints(uint256[] calldata _requestIds, uint256 _firstIndex, uint256 _lastIndex) external view returns (uint256[] memory hintIds) {
+        hintIds = new uint256[](_requestIds.length);
+        for (uint256 i = 0; i < _requestIds.length; ++i) {
+            hintIds[i] = findClaimHint(_requestIds[i], _firstIndex, _lastIndex);
```

Comment:
if order is enforced, we can start the search from the previously found hint

Reply:
Yep, it seems to be more intuitive. Agreed.